src/github.ts

```diff
@@ -146,12 +148,103 @@ export class Github implements GithubApi {
+  public async getCommit(sha: string) {
+    const commit = this.#octokit.rest.repos.getCommit({
+      ...this.getRepo(),
+      ref: sha,
+    });
+    return commit;
+  }
+
+  /**
+   * Retrieves the parent commit SHA of a given commit.
+   * If the commit is a merge commit, it returns the SHA of the first parent commit.
+   * @param sha - The SHA of the commit.
+   * @returns The SHA of the parent commit.
+   */
+  public async getParent(sha: string) {
+    const commit = await this.getCommit(sha);
+    // a commit has a parent. If it has more than one parent it is an indication
+    // that it is a merge commit. The first parent is the commit that was merged
+    // we can safely ignore the second parent as we're checking if the commit isn't a
+    // merge commit before.
+    return commit.data.parents[0].sha;
+  }
+
+  /**
+   * Retrieves the pull requests associated with a specific commit.
+   * @param sha The SHA of the commit.
+   * @returns A promise that resolves to the pull requests associated with the commit.
+   */
+  public async getPullRequestsAssociatedWithCommit(sha: string) {
+    const pr = this.#octokit.rest.repos.listPullRequestsAssociatedWithCommit({
+      ...this.getRepo(),
+      commit_sha: sha,
+    });
+    return pr;
+  }
+
+  /**
+   * Checks if a given SHA is associated with a specific pull request.
+   * @param sha - The SHA of the commit.
+   * @param pull - The pull request to check against.
+   * @returns A boolean indicating whether the SHA is associated with the pull request.
+   */
+  public async isShaAssociatedWithPullRequest(sha: string, pull: PullRequest) {
+    const assoc_pr = await this.getPullRequestsAssociatedWithCommit(sha);
+    const assoc_pr_data = assoc_pr.data;
+    // commits can be associated with multiple PRs
+    // checks if any of the assoc_prs is the same as the pull
+    return assoc_pr_data.some((pr) => pr.number == pull.number);
+  }
+
+  /**
+   * Checks if a pull request is "squashed and merged"
+   * or "rebased and merged"
+   * @param pull - The pull request to check.
+   * @returns A promise that resolves to a boolean indicating whether the pull request is squashed and merged.
+   */
+  public async isSquashed(pull: PullRequest): Promise<boolean> {
+    const merge_commit_sha = await this.getMergeCommitSha(pull);
+    if (!merge_commit_sha) {
+      console.log("likely not merged yet.");
+      return false;
+    }
+    // To detect if this was a rebase and merge, we can verify
+    // that the parent of the merge commit is associated with the pull request
+    // if it is, we have a "rebase and merge".
+    // if it is not, we have a "squash and merge".
+    const parent_commit = await this.getParent(merge_commit_sha);
+    const is_associated =
+      (await this.isShaAssociatedWithPullRequest(parent_commit, pull)) &&
+      (await this.isShaAssociatedWithPullRequest(merge_commit_sha, pull));
+    if (is_associated) {
+      return false;
+    }
+    return true;
+  }
```

Comment:
Thanks for the detailed analysis!

> 1️⃣ If the first parent of the merge commit is associated with the pull request, we know we're dealing with a multi-commit rebase. We can continue to traverse (and select) the ancestors of this commit until we encounter a commit that is not associated with the pull request. We can defensively assert that the number of detected rebased commits is equal to the number of commits on the pull request.

This can potentially reach the API's rate-limit very soon when the PR has a large number of commits as you'd have to send a `listPullRequestsAssociatedWithCommit` for every commit. Can't we boil down the logic to the following.

```
squashed:
  - > has single parent
  - > the parent of the `merge_commit_sha` is NOT associated with the PR

  => use merge_commit_sha

rebased single-commit:
  - > has single parent
  - > the parent of the `merge_commit_sha` is associated with the PR

  => use commits associated with the PR

rebased multi-commit:
  - > has single parent,
  - > the parent of the `merge_commit_sha` is associated with the PR

  => use commits associated with the PR

merge-commit:
  - > has multiple parents
  - > the parent of the `merge_commit_sha` is NOT associated with the PR

  => use commits associated with the PR

```

That would reduce the number of API calls to Github and keeps close to the original behavior. WDYT @korthout ?

Reply:
something like this -> https://github.com/korthout/backport-action/pull/399/commits/f626f14b5c254443879629a7e5870c3cf9388bd6